File: examples/src/org/apache/click/examples/page/ajax/scroller/AjaxLiveScroller.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.click.examples.page.ajax.scroller;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.annotation.Resource;
import org.apache.click.ActionResult;

import org.apache.click.element.Element;
import org.apache.click.element.JsImport;
import org.apache.click.element.JsScript;
import org.apache.click.examples.domain.Customer;
import org.apache.click.examples.page.BorderPage;
import org.apache.click.examples.service.CustomerService;
import org.apache.click.util.ClickUtils;
import org.apache.commons.lang.math.NumberUtils;
import org.springframework.stereotype.Component;

/**
 * Demonstrates how to to dynamically load more customer while scrolling down
 * the Page. An Ajax request is made to a pageAction that sends back the new
 * customers to display.
 */
@Component
public class AjaxLiveScroller extends BorderPage {

    private static final long serialVersionUID = 1L;

    @Resource(name="customerService")
    private CustomerService customerService;

    // Specifies the number of customers to retrieve at a time
    private int pageSize = 10;

    @Override
    public void onInit() {
        super.onInit();
        int offset = 0;
        List<Customer> customers = loadCustomers(offset, pageSize);
        addModel("customers", customers);
    }

    /**
     * A pageAction method that is called when the user scrolled to the bottom
     * of the page and more data is needed.
     */
    public ActionResult onScroll() {
        // Check if the offset parameter was received.
        int offset = NumberUtils.toInt(getContext().getRequest().getParameter("offset"));

        List<Customer> customers = loadCustomers(offset, pageSize);

        // Return customers between the given offset and pageSize
        Map actionResultModel = new HashMap();
        actionResultModel.put("customers", customers);
        actionResultModel.put("format", getFormat());

        // Return an action result for the customers.htm template and customers
        ActionResult actionResult = new ActionResult("/ajax/scroller/customers.htm", actionResultModel, ActionResult.HTML);
        return actionResult;
    }

    @Override
    public List<Element> getHeadElements() {
        // Lazily load head elements and ensure they are only loaded once
        if (headElements == null) {
            headElements = super.getHeadElements();

            // Add the jQuery library
            headElements.add(new JsImport("/assets/js/jquery-1.4.2.js"));

            // Create a default model and add the pageSize variable to pass to
            // the JavaScript template: ajax-live-scroller.js
            Map<String, Object> jsModel = ClickUtils.createTemplateModel(this, getContext());
            jsModel.put("pageSize", pageSize);

            // Note the actual JavaScript necessary to setup the dynamic scrolling
            // is specified in the Page JavaScript template: ajax-live-scroller.js.
            headElements.add(new JsScript("/ajax/scroller/ajax-live-scroller.js", jsModel));
        }
        return headElements;
    }

    private List<Customer> loadCustomers(int offset, int pageSize) {
        return customerService.getTopCustomersForPage(offset, pageSize);
    }
}

File: examples/webapp/ajax/scroller/ajax-live-scroller.js
// #*
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
// *#

$(document).ready(function(){
    // Dynamically create a fixed scroll indicator at the bottom right corner of the screen
    jQuery("<div id='scroll-indicator' style='border: 1px solid navy;"
        + "background: #334ECF; color: #FFFFFF; padding: 10px; position: fixed; display: none;"
        + "right: 25px; bottom: 25px; width: 75px;"
        + "'>Loading...</div>")
        .insertAfter(jQuery('div.page')); // Insert the indicator after the main div defined in border-template.htm

    // Current number of records loaded from server
    var offset = 0;

    // Indicates the number of records to return from server. The PageSize value
    // is passed in from the AjaxLiveScroller.java Page.
    var pageSize = $pageSize;

    // Flag indicating if data is being loaded from server
    var loading = false;

    function loadMoreData() {
        // Toggle flag to indicate that loading started
        loading = true;

        // Increment offset with the specified number of pages
        offset+=pageSize;

        // Show the scroll indicator
        jQuery('#scroll-indicator').css("display", "block");

        // Perform Ajax GET request. Note pageAction parameter which specifies the onScroll page method
        jQuery.get('${context}${path}?pageAction=onScroll&offset=' + offset, function(data){
            // If the server response contains no div element, unbind the window scroll event as no further data
            // will be returned
            if (data.indexOf("<div") < 0) {
                jQuery(window).unbind('scroll');
            } else {
                // Otherwise we add the new data after the last customer
                jQuery(".customer:last").after(data);
            }

            jQuery('#scroll-indicator').css("display", "none");

            // Toggle flag to indicate that loading is finished
            loading = false;
        });
    }

    // Bind a scroll event to the window to detect if more data needs to be loaded
    jQuery(window).bind('scroll', function() {

        // If we are currently retrieving data, don't load again
        if (loading) {
            return;
        }

        // When scrolling near the bottom of the page, load more data from server
        if (nearEndOfPage()) {
            // Incase we scrolled to end of page, move scrollbar slightly back
            jQuery(window).scrollTop(jQuery(window).scrollTop() - 1);
            loadMoreData();
        }
    });

    // Return true if scrolling is 85% at the end of the page
    function nearEndOfPage() {
        return jQuery(window).scrollTop() > getFetchThreshold();
    }

    // Return the threshold when more data should be fetched, equates to 85% of the document height
    function getFetchThreshold() {
        return 0.85 * jQuery(document).height() - jQuery(window).height();
    }
});
